pkg/util/ranger/types.go

```diff
@@ -282,6 +282,71 @@ func formatDatum(d types.Datum, isLeftSide bool) string {
 	return fmt.Sprintf("%v", d.GetValue())
 }
 
+// compareLexicographically compares two bounds from two ranges and returns 0, 1, -1
+// for equal, greater than or less than respectively. It gets the two bounds,
+// collations and if each bound is open (open1, open2) or closed. In addition,
+// it also gets if each bound is lower or upper (low1, low2).
+// Lower bounds logically can be extended with -infinity and upper bounds can be extended with +infinity.
+func compareLexicographically(tc types.Context, bound1, bound2 []types.Datum, collators []collate.Collator,
+	open1, open2, low1, low2 bool) (int, error) {
+	n1 := len(bound1)
+	n2 := len(bound2)
+	n := min(n1, n2)
+
+	for i := 0; i < n; i++ {
+		cmp, err := bound1[i].Compare(tc, &bound2[i], collators[i])
+		if err != nil {
+			return 0, err
+		}
+		if cmp != 0 {
+			return cmp, nil
+		}
+	}
+
+	// Handle interval types
+	if n1 == n2 {
+		switch {
+		case !open1 && !open2:
+			return 0, nil
+		case open1 == open2:
+			if low1 == low2 {
+				return 0, nil
+			} else if low1 {
+				return 1, nil
+			} else {
+				return -1, nil
+			}
+		case open1:
+			if low1 {
+				return 1, nil
+			}
+			return -1, nil
+		case open2:
+			if low2 {
+				return -1, nil
+			}
+			return 1, nil
+		}
+	}
+
+	// Unequal length ranges. We use -infinity for lower bounds and +infinity for upper bounds.
+	if n1 < n2 {
+		if low1 {
+			// -infinity is less than anything
+			return -1, nil
+		}
+		// +infinity is higher than anything
+		return 1, nil
```

Comment:
It seems different `open1` should result in different results.

Reply:
it will be the same here if it is open or close because it is -infinity. -infinity is lower than any other value.

Reply:
It becomes a bit complicated here and I think it relates to the definition of what a range with fewer columns means.
I come up with one example: `[1 1,1 5) [1 5 1,1 6 5]`. I think `1 5)` should be less than `[1 5 1` because the first range should come from expressions like `a = 1 and b >= 1 and b < 5`.

Reply:
Looks like the key issue here is that does `[1 1,1 5)` mean `a = 1 and b >= 1 and b < 5` or `a = 1 and b >= 1 and b < 6`.

Reply:
The example above is not applicable, You are using a general n dimensional range and not the lexiographically ordered index. [1 5 1,1 6 5] is not a valid index range. All the prefix should be single point. The open and closed is applied to all the columns. So do you mean (1 5 compared to [1 5 1, yes it is less than and consistent with the code.

Reply:
[1 1,1 5) mean a = 1 and b >= 1 and b < 5

Reply:
> [1 5 1,1 6 5] is not a valid index range

This can be a valid range. If the user writes`(a,b,c) >= (1,5,1) and (a,b,c) <= (1,6,5)`, that's exactly what we want.

> If you mean comparing the upper bound 1 5) to lower bound [1 5 1 then you are comparing 1 5 +infinity) and [1 5 1 and 1 5 +infinity) is higher.

If `[1 1,1 5)` means `a = 1 and b >= 1 and b < 5`, then `1 5)` means `1 5 -infinity)` instead of `1 5 +infinity)`, right? Because `a = 1 and b = 5 and c = -999` doesn't satisfy `a = 1 and b >= 1 and b < 5`.
Then `1 5)` should be lower than `[1 5 1`.

Reply:
1 5) as an upper bound means a=1 and b < 5 and c anything which means c < +infinity. If you take a simple example of [1 5, 1 5] it means a=1 and b=1 and all values of c which means c >= -infinity for low range and c <= +infinity for the upper bound.